File: cpp/Platform.Collections.Methods/RecursionlessSizeBalancedTreeMethods.cpp
namespace Platform::Collections::Methods::Trees
{
    template <typename TElement> class RecursionlessSizeBalancedTreeMethods : public SizedBinaryTreeMethodsBase<TElement>
    {
    public:
        void AttachCore(TElement* root, TElement node) override
        {
            while (true)
            {
                auto* left = this->GetLeftReference(*root);
                auto leftSize = this->GetSizeOrZero(*left);
                auto* right = this->GetRightReference(*root);
                auto rightSize = this->GetSizeOrZero(*right);
                if (this->FirstIsToTheLeftOfSecond(node, *root))
                {
                    if (*left == 0)
                    {
                        this->IncrementSize(*root);
                        this->SetSize(node, 1);
                        *left = node;
                        return;
                    }
                    if (this->FirstIsToTheLeftOfSecond(node, *left))
                    {
                        if ((leftSize + 1) > rightSize)
                        {
                            this->RightRotate(root);
                        }
                        else
                        {
                            this->IncrementSize(*root);
                            root = left;
                        }
                    }
                    else
                    {
                        auto leftRightSize = this->GetSizeOrZero(this->GetRight(*left));
                        if ((leftRightSize + 1) > rightSize)
                        {
                            if (leftRightSize == 0 && rightSize == 0)
                            {
                                this->SetLeft(node, *left);
                                this->SetRight(node, *root);
                                this->SetSize(node, leftSize + 2);
                                this->SetLeft(*root, 0);
                                this->SetSize(*root, 1);
                                *root = node;
                                return;
                            }
                            this->LeftRotate(left);
                            this->RightRotate(root);
                        }
                        else
                        {
                            this->IncrementSize(*root);
                            root = left;
                        }
                    }
                }
                else
                {
                    if (*right == 0)
                    {
                        this->IncrementSize(*root);
                        this->SetSize(node, 1);
                        *right = node;
                        return;
                    }
                    if (this->FirstIsToTheRightOfSecond(node, *right))
                    {
                        if ((rightSize + 1) > leftSize)
                        {
                            this->LeftRotate(root);
                        }
                        else
                        {
                            this->IncrementSize(*root);
                            root = right;
                        }
                    }
                    else
                    {
                        auto rightLeftSize = this->GetSizeOrZero(this->GetLeft(*right));
                        if ((rightLeftSize + 1) > leftSize)
                        {
                            if (rightLeftSize == 0 && leftSize == 0)
                            {
                                this->SetLeft(node, *root);
                                this->SetRight(node, *right);
                                this->SetSize(node, rightSize + 2);
                                this->SetRight(*root, 0);
                                this->SetSize(*root, 1);
                                *root = node;
                                return;
                            }
                            this->RightRotate(right);
                            this->LeftRotate(root);
                        }
                        else
                        {
                            this->IncrementSize(*root);
                            root = right;
                        }
                    }
                }
            }
        }

        void DetachCore(TElement* root, TElement node) override
        {
            while (true)
            {
                auto* left = this->GetLeftReference(*root);
                auto leftSize = this->GetSizeOrZero(*left);
                auto* right = this->GetRightReference(*root);
                auto rightSize = this->GetSizeOrZero(*right);
                if (this->FirstIsToTheLeftOfSecond(node, *root))
                {
                    auto decrementedLeftSize = leftSize - 1;
                    if (this->GetSizeOrZero(this->GetRight(*right)) > decrementedLeftSize)
                    {
                        this->LeftRotate(root);
                    }
                    else if (this->GetSizeOrZero(this->GetLeft(*right)) > decrementedLeftSize)
                    {
                        this->RightRotate(right);
                        this->LeftRotate(root);
                    }
                    else
                    {
                        this->DecrementSize(*root);
                        root = left;
                    }
                }
                else if (this->FirstIsToTheRightOfSecond(node, *root))
                {
                    auto decrementedRightSize = rightSize - 1;
                    if (this->GetSizeOrZero(this->GetLeft(*left)) > decrementedRightSize)
                    {
                        this->RightRotate(root);
                    }
                    else if (this->GetSizeOrZero(this->GetRight(*left)) > decrementedRightSize)
                    {
                        this->LeftRotate(left);
                        this->RightRotate(root);
                    }
                    else
                    {
                        this->DecrementSize(*root);
                        root = right;
                    }
                }
                else
                {
                    if (leftSize > 0 && rightSize > 0)
                    {
                        TElement replacement = 0;
                        if (leftSize > rightSize)
                        {
                            replacement = this->GetRightest(*left);
                            this->DetachCore(left, replacement);
                        }
                        else
                        {
                            replacement = this->GetLeftest(*right);
                            this->DetachCore(right, replacement);
                        }
                        this->SetLeft(replacement, *left);
                        this->SetRight(replacement, *right);
                        this->SetSize(replacement, leftSize + rightSize);
                        *root = replacement;
                    }
                    else if (leftSize > 0)
                    {
                        *root = *left;
                    }
                    else if (rightSize > 0)
                    {
                        *root = *right;
                    }
                    else
                    {
                        *root = 0;
                    }
                    this->ClearNode(node);
                    return;
                }
            }
        }
    };
}

File: csharp/Platform.Collections.Methods/Trees/RecursionlessSizeBalancedTreeMethods.cs
using System.Numerics;

#pragma warning disable CS1591 // Missing XML comment for publicly visible type or member

namespace Platform.Collections.Methods.Trees
{
    /// <summary>
    /// <para>
    /// Represents the recursionless size balanced tree methods.
    /// </para>
    /// <para></para>
    /// </summary>
    /// <seealso cref="SizedBinaryTreeMethodsBase{TElement}"/>
    public abstract class RecursionlessSizeBalancedTreeMethods<TElement> : SizedBinaryTreeMethodsBase<TElement> where TElement: IUnsignedNumber<TElement>, IComparisonOperators<TElement, TElement, bool>
    {
        /// <summary>
        /// <para>
        /// Attaches the core using the specified root.
        /// </para>
        /// <para></para>
        /// </summary>
        /// <param name="root">
        /// <para>The root.</para>
        /// <para></para>
        /// </param>
        /// <param name="node">
        /// <para>The node.</para>
        /// <para></para>
        /// </param>
        protected override void AttachCore(ref TElement root, TElement node)
        {
            while (true)
            {
                ref var left = ref GetLeftReference(root);
                var leftSize = GetSizeOrZero(left);
                ref var right = ref GetRightReference(root);
                var rightSize = GetSizeOrZero(right);
                if (node < root) // node.Key less than root.Key
                {
                    if (left == TElement.Zero)
                    {
                        IncrementSize(root);
                        SetSize(node, TElement.One);
                        left = node;
                        return;
                    }
                    if (node < left) // node.Key less than left.Key
                    {
                        if ((leftSize + TElement.One) > rightSize)
                        {
                            RightRotate(ref root);
                        }
                        else
                        {
                            IncrementSize(root);
                            root = ref left;
                        }
                    }
                    else  // node.Key greater than left.Key
                    {
                        var leftRightSize = GetSizeOrZero(GetRight(left));
                        if ((leftRightSize + TElement.One) > rightSize)
                        {
                            if (leftRightSize == TElement.Zero && rightSize == TElement.Zero)
                            {
                                SetLeft(node, left);
                                SetRight(node, root);
                                SetSize(node, (leftSize+ TElement.CreateTruncating(2))); // TElement.CreateTruncating(2) (2) - node the size of root and a node itself
                                SetLeft(root, TElement.Zero);
                                SetSize(root, TElement.One);
                                root = node;
                                return;
                            }
                            LeftRotate(ref left);
                            RightRotate(ref root);
                        }
                        else
                        {
                            IncrementSize(root);
                            root = ref left;
                        }
                    }
                }
                else // node.Key greater than root.Key
                {
                    if (right == TElement.Zero)
                    {
                        IncrementSize(root);
                        SetSize(node, TElement.One);
                        right = node;
                        return;
                    }
                    if (node > right) // node.Key greater than right.Key
                    {
                        if ((rightSize + TElement.One) > leftSize)
                        {
                            LeftRotate(ref root);
                        }
                        else
                        {
                            IncrementSize(root);
                            root = ref right;
                        }
                    }
                    else // node.Key less than right.Key
                    {
                        var rightLeftSize = GetSizeOrZero(GetLeft(right));
                        if ((rightLeftSize + TElement.One) > leftSize)
                        {
                            if (rightLeftSize == TElement.Zero && leftSize == TElement.Zero)
                            {
                                SetLeft(node, root);
                                SetRight(node, right);
                                SetSize(node, (rightSize + TElement.CreateTruncating(2))); // TElement.CreateTruncating(2) (2) - node the size of root and a node itself
                                SetRight(root, TElement.Zero);
                                SetSize(root, TElement.One);
                                root = node;
                                return;
                            }
                            RightRotate(ref right);
                            LeftRotate(ref root);
                        }
                        else
                        {
                            IncrementSize(root);
                            root = ref right;
                        }
                    }
                }
            }
        }

        /// <summary>
        /// <para>
        /// Detaches the core using the specified root.
        /// </para>
        /// <para></para>
        /// </summary>
        /// <param name="root">
        /// <para>The root.</para>
        /// <para></para>
        /// </param>
        /// <param name="node">
        /// <para>The node.</para>
        /// <para></para>
        /// </param>
        protected override void DetachCore(ref TElement root, TElement node)
        {
            while (true)
            {
                ref var left = ref GetLeftReference(root);
                var leftSize = GetSizeOrZero(left);
                ref var right = ref GetRightReference(root);
                var rightSize = GetSizeOrZero(right);
                if (node < root) // node.Key less than root.Key
                {
                    var decrementedLeftSize = leftSize - TElement.One;
                    if ((GetSizeOrZero(GetRightOrDefault(right))) > decrementedLeftSize)
                    {
                        LeftRotate(ref root);
                    }
                    else if ((GetSizeOrZero(GetLeftOrDefault(right))) > decrementedLeftSize)
                    {
                        RightRotate(ref right);
                        LeftRotate(ref root);
                    }
                    else
                    {
                        DecrementSize(root);
                        root = ref left;
                    }
                }
                else if (node > root) // node.Key greater than root.Key
                {
                    var decrementedRightSize = rightSize - TElement.One;
                    if ((GetSizeOrZero(GetLeftOrDefault(left))) > decrementedRightSize)
                    {
                        RightRotate(ref root);
                    }
                    else if ((GetSizeOrZero(GetRightOrDefault(left))) > decrementedRightSize)
                    {
                        LeftRotate(ref left);
                        RightRotate(ref root);
                    }
                    else
                    {
                        DecrementSize(root);
                        root = ref right;
                    }
                }
                else // key equals to root.Key
                {
                    if (leftSize > TElement.Zero && rightSize > TElement.Zero)
                    {
                        TElement replacement;
                        if ((leftSize) > rightSize)
                        {
                            replacement = GetRightest(left);
                            DetachCore(ref left, replacement);
                        }
                        else
                        {
                            replacement = GetLeftest(right);
                            DetachCore(ref right, replacement);
                        }
                        SetLeft(replacement, left);
                        SetRight(replacement, right);
                        SetSize(replacement, (leftSize + rightSize));
                        root = replacement;
                    }
                    else if (leftSize > TElement.Zero)
                    {
                        root = left;
                    }
                    else if (rightSize > TElement.Zero)
                    {
                        root = right;
                    }
                    else
                    {
                        root = TElement.Zero;
                    }
                    ClearNode(node);
                    return;
                }
            }
        }
    }
}
